sr/robot3/astoria.py

```diff
@@ -28,126 +31,197 @@ class GetMetadataResult(NamedTuple):
-        """Use the logging from sr-robot3."""
-        # Suppress INFO messages from gmqtt
-        logging.getLogger("gmqtt").setLevel(logging.WARNING)
+    def __init__(self, mqtt_client: MQTTClient) -> None:
+        self._mqtt = mqtt_client
 
-    def _init(self) -> None:
-        """Initialise consumer."""
+        # Initialise consumer data structures.
         self._metadata_message: Optional[MetadataManagerMessage] = None
         self._proc_message: Optional[ProcessManagerMessage] = None
-        self._state_lock = asyncio.Lock()
+        self._astmetad_received = Event()
+        self._astprocd_received = Event()
+
+    def run(self, timeout: Optional[float] = None) -> bool:
+        """Entrypoint for the data component."""
         self._mqtt.subscribe("astmetad", self._handle_astmetad_message)
         self._mqtt.subscribe("astprocd", self._handle_astprocd_message)
 
-    async def _handle_astmetad_message(
+        timeout_step = (timeout / 2) if timeout is not None else None
+
+        # wait for both messages to be received with a timeout
+        # the division of time between each step is irrelevant,
+        # messages can arrive for either during the whole period
+        self._astmetad_received.wait(timeout=timeout_step)
+        metadata_received = self._astprocd_received.wait(timeout=timeout_step)
+
+        self._mqtt.unsubscribe("astmetad")
+        self._mqtt.unsubscribe("astprocd")
+
+        return metadata_received
+
+    def _handle_astmetad_message(
         self,
-        match: Match[str],
-        payload: str,
+        client: MQTT,
+        userdata: Any,
+        msg: MQTTMessage,
     ) -> None:
         """Handle astmetad status messages."""
-        async with self._state_lock:
-            try:
-                message = parse_obj_as(MetadataManagerMessage, loads(payload))
-                if message.status == MetadataManagerMessage.Status.RUNNING:
-                    LOGGER.debug("Received metadata")
-                    self._metadata_message = message
-                else:
-                    LOGGER.warn("Cannot get metadata, astmetad is not running")
-            except JSONDecodeError:
-                LOGGER.error("Could not decode JSON metadata.")
-            if self._metadata_message is not None and self._proc_message is not None:
-                self.halt(silent=True)
-
-    async def _handle_astprocd_message(
+        try:
+            message = parse_obj_as(MetadataManagerMessage, loads(msg.payload))
+            if message.status == MetadataManagerMessage.Status.RUNNING:
+                LOGGER.debug("Received metadata")
+                self._metadata_message = message
+            else:
+                LOGGER.warn("Cannot get metadata, astmetad is not running")
+        except JSONDecodeError:
+            LOGGER.error("Could not decode JSON metadata.")
+        if self._metadata_message is not None:
+            self._astmetad_received.set()
```

Comment:
It's not possible for `message` to be `None` here. Pydantic will either return a parsed and valid `MetadataManagerMessage` OR throw a `pydantic.ValidationError`.

```suggestion
        try:
            message = parse_obj_as(MetadataManagerMessage, loads(msg.payload))
            if message.status == MetadataManagerMessage.Status.RUNNING:
                LOGGER.debug("Received metadata")
                self._metadata_message = message
                self._astmetad_received.set()
            else:
                LOGGER.warn("Cannot get metadata, astmetad is not running")
        except JSONDecodeError:
            LOGGER.error("Could not decode JSON metadata.")
        except pydantic.ValidationError as e:
            LOGGER.error(f"Unable to parse metadata message: {e}")
            
```